### ISSUE TITLE
Harden ShardCoordinator state update guarantees

### ISSUE THREAD

**Issue by patriknw** [akka/akka#28856]
A few questions were raised about if the ShardCoordinator state write/read with majority and min-cap=5 is enough. Problem is when cluster size changes between the write and the read.

Scenario 1: For small clusters it should be fine as long as you don't remove more nodes than majority-min-cap at the same time. Here is a scenario that could be problem.
* 7 nodes, N1..N7. Coordinator on N1 decides that a shard should be hosted by N7.
* That is updated and it's started. That update is replicated to N1..N5 but not to N6 and N7 yet.
* All N1..N5 are shutdown and new coordinator is started on N6.
* New coordinator doesn't know that the shard is running on N7 and makes a new decision to host it on N6.
* Then it's running on both N6 and N7.

Scenario 2: For larger clusters the update is immediately replicated to more nodes (majority) before starting the shard. It can still be a theoretical problem even though the risk is lower.
* For example, 99 nodes, N1..N99. Coordinator on N1 decides that a shard should be hosted by N99.
* That is updated and it's started. That update is replicated to N1..N50 but not to N51 and N99 yet.
* 10 nodes, N1..N10 are shutdown and new coordinator is started on N51.
* New coordinator is only required to read from 45 nodes and by chance only reads from N51..N96 (not from N11..N50) so it doesn't know that the shard is running on N99. Makes a new decision to host it on N98.
* Then it's running on both N98 and N99.

Although these edge cases seems very unlikely we should think one more time about them and see if we can harden it further.

Maybe the regions can include their shards when registering to new coordinator, but that would require that coordinator can't make any new decisions until all expected regions have registered?

**Comment by patriknw** [comment]
Here is another scenario that isn't covered fully:

* C1 only succeeds in writing to a minority of nodes, so the shard isn't started yet
* C1 is shutdown
* C2 is started and sees the update from C1, and therefore starts the shard
* C2 is shutdown
* C3 is started, doesn't see the update from C1 (still only on some minority of nodes)
* C3 decides to start the shard elsewhere

A solution to that could be that C2 must write to majority also for the initial state before starting such shards.

**Comment by dricketts** [comment]
There might be a related scenario caused by the timestamp logic in LWWRegister (assuming it uses the defaultClock described in [this page](https://doc.akka.io/api/akka/2.6.4/akka/cluster/ddata/LWWRegister.html)):

1. C1 starts on a node with a wall clock far ahead of any other node in the system.
2. C1 writes an update to a minority of nodes. This update could be a re-replication of the existing state (if the above solution is implemented) or a new update.
3. C1 is shutdown.
4. C2 is started and does not see the timestamp of the update from C1.
5. C2 makes some changes and updates the state on a majority of nodes. However, this update has a timestamp less than the minority-replicated update from C1.
6. Through gossip, the update from C1 spreads and overwrites the update from C2.

Unless I'm missing something, there seems to inadequate fencing of updates from old coordinators.

**Comment by patriknw** [comment]
Looks like all these cases boil down to that the coordinator should read from more nodes when starting up. It could use `ReadAll`, but can we use something less?

For scenario 4 with the clock skew we could maybe adjust the clock for the initial write that was suggested for scenario 3. It could advance the clock, let's say 10 seconds, for that write. Assuming decent clock synchronization.

For scenario 2, the large cluster, we could introduce something like `ReadMajority + n`, which would read from n more nodes than the majority count (up to all nodes). Configure n depending on how many nodes you need to be able to leave at the same time. If I'm calculating right n=5 would be enough for the scenario 2 when leaving 10 nodes.

For scenario 2 with crashed nodes, downing instead of leaving, I don't think it's much of a problem in practice, since failure detection, downing and removal takes rather long time. Giving ddata enough time to disseminate.

Scenario 1 is already covered by majority-min-cap, assuming that the downing provider would not keep the 2 nodes in minority described in that scenario.

**Comment by johanandren** [comment]
I'm thinking that we should just require syncing clocks and never having "a wall clock far ahead of any other node in the system". Is that unreasonable? (We should perhaps document that better though)

Some thoughts:
 * The singleton will only switch over to a new node triggered by leader action, wouldn't it be relatively safe to do a `ReadAll`? Perhaps we should provide both as options? 
 * Since we know the singleton will hand over to the new oldest node, could we explicitly do writes to the `n` oldest nodes somehow?
